### DISCUSSION TITLE
[Feat]: Improve performance by disable auto play when not visible

### DISCUSSION THREAD

**Discussion by njfamirm** [davidjerleke/embla-carousel#1010] [Ideas] [answered]
### Which variants of Embla Carousel are relevant to this feature request?

- [ ] embla-carousel **(Core)**
- [ ] embla-carousel-react
- [ ] embla-carousel-vue
- [ ] embla-carousel-svelte
- [X] embla-carousel-autoplay
- [ ] embla-carousel-auto-scroll
- [ ] embla-carousel-solid
- [ ] embla-carousel-auto-height
- [ ] embla-carousel-class-names
- [ ] embla-carousel-fade
- [ ] embla-carousel-docs **(Documentation)**
- [ ] embla-carousel-docs **(Generator)**

### Feature description

Hi

Using built-in browser APIs, we can detect when an element is not visible, preventing unnecessary autoplay and CPU-intensive calculations like opacity updater.

this is my solution:
```ts
function stopAutoPlayOnNotVisible(carouselNode: HTMLElement, emblaApi: EmblaCarouselType): void {
  logger.logMethod?.('stopAutoPlayOnNotVisible');

  const autoPlayPlugin = emblaApi.plugins().autoplay;

  const observer = new IntersectionObserver((entries) => {
    const entry = entries[0];
    if (entry.isIntersecting) {
      logger.logMethod?.('autoPlayPlugin.play');
      autoPlayPlugin.play();
    }
    else {
      logger.logMethod?.('autoPlayPlugin.stop');
      autoPlayPlugin.stop();
    }
  });

  observer.observe(carouselNode);
}
```

### Additional Context

Additional details here...

### Before submitting

- [X] I've made research efforts and searched the documentation
- [X] I've searched for existing issues
- [X] I agree to follow this project's Contributing Guidelines for feature requests

**Comment by davidjerleke** [comment]
@njfamirm is this a question or feature request? Do you want the in view feature to be built-in for the Autoplay plugin?

**Reply by njfamirm** [reply]
`feature request`, yes if your ok with this

**Reply by davidjerleke** [reply]
@njfamirm have you tried this?

```tsx
emblaApi.on('slidesInView', (emblaApi) => {
  const autoplay = emblaApi.plugins().autoplay;
  if (!autoplay) return;

  const hasSlidesInView = emblaApi.slidesInView().length > 0;

  if (hasSlidesInView) {
    autoplay.play()
  } else {
    autoplay.stop()
  }
});
```

**Reply by davidjerleke** [reply]
@njfamirm have you seen [my response](https://github.com/davidjerleke/embla-carousel/discussions/1010#discussioncomment-11371527)?

**Reply by wonu** [reply]
I think @njfamirm might be referring to whether the `root` node itself is intersecting with the viewport. It seems `slidesInView` is determined based on its position relative to `container.parentElement`, which wouldn't necessarily reflect if it's actually intersecting with the main viewport, right?

I'm new to this library (just started looking at it today), so I could be wrong, but this seems like a potentially workaround. It would be ideal if there were a built-in API for this:
```ts
useEffect(() => {
  if (emblaApi) {
    new IntersectionObserver(
      (entries, observer) => {
        if (entries[0].isIntersecting) {
          emblaApi.plugins().autoplay.play()
        } else {
          emblaApi.plugins().autoplay.stop()
        }
      },
      {
        threshold: 0.5,
      },
    ).observe(emblaApi.rootNode())
  }
}, [emblaApi])
```

**Reply by njfamirm** [reply]
I apologize for the delay in responding!

I'm test your (@davidjerleke) solution, but as @wonu said:

> the `root` node itself is intersecting with the viewport. It seems `slidesInView` is determined based on its position relative to `container.parentElement`, which wouldn't necessarily reflect if it's actually intersecting with the main viewport

Given the relatively high processing power, resolving this issue can prevent significant performance issues.

**Reply by davidjerleke** [reply, marked as answer]
> the `root` node itself is intersecting with the viewport. It seems `slidesInView` is determined based on its position relative to `container.parentElement`, which wouldn't necessarily reflect if it's actually intersecting with the main viewport

@wonu You're absolutely right—I must've been running on low battery when I put that together. That said, I generally expect most projects to include some kind of "in view" utility.

For React, something like `useInView()` would significantly reduce the boilerplate you're using with `IntersectionObserver` in [your snippet here](https://github.com/davidjerleke/embla-carousel/discussions/1010#discussioncomment-12783195).

I don't think this kind of functionality belongs in the plugin itself, though—it would add unnecessary weight. Embla’s core idea is to stay lightweight and flexible, giving you the tools to build your own tailored solutions.

If you're looking for a more fully featured setup, there are some great UI libraries that build on Embla and offer extra logic out of the box. A few worth checking out can be found here:

- https://github.com/davidjerleke/embla-carousel/discussions/691#discussion-6074607

**Reply by njfamirm** [reply]
> I don't think this kind of functionality belongs in the plugin itself, though—it would add unnecessary weight. Embla’s core idea is to stay lightweight and flexible, giving you the tools to build your own tailored solutions.

Well, considering the purpose of the project you mentioned, then I'll close this issue.
Thank you
